core/src/main/java/io/grpc/Metadata.java

```diff
@@ -283,8 +395,25 @@ public boolean containsKey(Key<?> key) {
    */
   @ExperimentalApi
   public <T> void discardAll(Key<T> key) {
```

Comment:
The JVM is not smart enough to skip creating the Iterable when it isn't used by the caller of removeAll.  (which is why ti showed up in the profile).  I expect that most of the time there is nothing to discard at all (we use it primarily for sanity checks) so I want it to be as fast as possible.